### PR TITLE
Non-recursive vm_loop

### PR DESCRIPTION
**Pull request by zherczeg**
Processing calls is removed from vm_loop which heavily improves the performance of recursive heavy benchmarks. E.g. controlflow-recursive memory consumption is reduced by 37%.

### CODE REVIEW

**Comment by zherczeg** [comment]
On RPi2, controlflow-recursive.js memory is reduced by 37.5%, and 3d-cube.js memory is reduced by 18.1%. Other tests are unaffected. Please check the patch.

File: jerry-core/vm/vm.c [hunk 1]
```diff
@@ -328,6 +328,128 @@ vm_get_implicit_this_value (ecma_value_t *this_value_p) /**< in-out: this value
...
 /**
+ * 'Function call' opcode handler.
+ *
+ * See also: ECMA-262 v5, 11.2.3
+ */
+static void
+opfunc_call (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+{
+  uint8_t opcode = frame_ctx_p->byte_code_p[0];
+  unsigned int arguments_list_len;
```
Comment (LaszloLango) [inline jerry-core/vm/vm.c]: I think we should use `uint32_t` or `uint16_t` instead.

File: jerry-core/vm/vm.c [hunk 2]
```diff
@@ -1216,50 +1338,22 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
         case VM_OC_CALL:
-        case VM_OC_CALL_PROP:
         {
-          ecma_value_t this_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+          JERRY_ASSERT (free_flags == 0);
 
-          stack_top_p -= right_value;
-
-          if (VM_OC_GROUP_GET_INDEX (opcode_data) >= VM_OC_CALL_PROP_N)
+          if (!frame_ctx_p->call_operation)
```
Comment (LaszloLango) [inline jerry-core/vm/vm.c]: This is a bit hard to read. Is this condition the same as `frame_ctx_p->call_operation == VM_NO_EXEC_OP`?

File: jerry-core/vm/vm.c [hunk 3]
```diff
@@ -328,6 +328,128 @@ vm_get_implicit_this_value (ecma_value_t *this_value_p) /**< in-out: this value
 } /* vm_get_implicit_this_value */
 
 /**
+ * 'Function call' opcode handler.
+ *
+ * See also: ECMA-262 v5, 11.2.3
+ */
+static void
+opfunc_call (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
```
Comment (akiss77) [inline jerry-core/vm/vm.c]: Why are `opfunc_*` functions in vm.c? Don't all such functions belong to opcodes.c?

File: jerry-core/vm/vm.h [hunk 4]
```diff
@@ -184,13 +178,21 @@ typedef enum
   VM_OC_PUT_BLOCK = VM_OC_PUT_DATA_CREATE_FLAG (0x8),
 } vm_oc_put_types;
 
+/**
+ * Result writers that are part of the opcodes.
```
Comment (akiss77) [inline jerry-core/vm/vm.h]: Please rephrase this to be more descriptive.

File: jerry-core/vm/vm.c (hunk 3 above)
Comment (zherczeg) [inline jerry-core/vm/vm.c]: Some opfuncs are part of vm.c because opcodes.c does not depend on byte code format. We could refactor this but I would do it in a seperate patch.

**Comment by zherczeg** [comment]
Patch updated.

**Comment by LaszloLango** [comment]
LGTM

**Comment by zherczeg** [comment]
self LGTM

**Comment by zherczeg** [comment]
Landed in 1200be42b4cce8bfa9f2a7871d1df26b9aea05f4
